core/dbt/events/functions.py

```diff
@@ -22,57 +22,65 @@ def scrub_secrets(msg: str, secrets: List[str]) -> str:
     return scrubbed
 
 
+# this exists because some log messages are actually expensive to build.
+# for example, many debug messages call `dump_graph()` and we don't want to
+# do that in the event that those messages are never going to be sent to
+# the user because we are only logging info-level events.
+def lazy_log_msg(e: CliEventABC) -> Callable[[], str]:
+    return lambda: scrub_secrets(e.cli_msg(), env_secrets())
+
```

Comment:
This looks great @nathaniel-may!  One thought, could we move `logger.timestamped_line` into this?  it would dry up the big route-all-the-things block below.

Reply:
oh good call. `timestamped_line` actually shouldn't be called at all because structlog is adding those now. I'll drop that in #4176.

Reply:
jk I threw it in here.